### ISSUE TITLE
Current state of support on ARM platforms

### ISSUE THREAD

**Issue by vtk9** [gperftools/gperftools#841]
I am wondering what the current status of support for ARM architecture is (specifically armv7 raspberry pi running ubuntu 16.04) From my research on the web and this github project community, I've learned that Gperftools memory profiling on ARM is not as robust as on x86. 

I've had some success on ARM using libunwind with very simple applications to verify that memory profiling is possible, but as soon as I attempt to instrument a complex application, it crashes somewhere in libunwind. With all the different ways of capturing the stacktrace (libunwind, libgcc unwind etc.) validation of the unit tests via make check results with, at best, 8 failures. Most of these failures happen in the heap-checker tests.

The discussion in https://github.com/gperftools/gperftools/issues/751 (closed Feb 21, 2016) seems to indicate that memory profiler seems to work with liggcc-based unwinder. I have been unable to get it to work. Like the author in issue #751, the heap info is generated but it is missing the information. Is there anything else I am missing? Are there any other flags or parameters I need to pass other than

--enable-stacktrace-via-backtrace --enable-emergency-malloc 

What is the recommended way to use the heap-profiler feature on ARM, if any?

**Comment by alk** [comment]
Yes, unfortunately libunwind doesn't seem to be 100% supported on less major platforms. That's why not linking with libunwind and using, say, libgcc should work better.

Keep in mind however, that a) this won't work with signal handlers and thus with cpu profiler b) all dwarf-based unwinders (and whatever ARM uses are similar) are super slow. Which makes them nearly useless for stuff that is doing backtrace capturing often (e.g. heap profiler).

As for your problem specifically, keep in mind that libunwind does interpose it's own backtrace() implementation over glibc's. So if you want to use glibc's backtrace that uses libgcc, you need to stop linking with libunwind.

Ideal solution for quick backtrace support for arm would be frame-pointer-based asm unwinder. But that requires someone contributing one.

**Comment by vtk9** [comment]
Thank you for your response. 

Sounds like at this point, libgcc's backtrace is the only option left. I have not been able to profile on ARM with libgcc's stacktrace. Specifically, I am using the gperftools API and calling HeapProfilerDump(...).. but the snapshot that gets created does not have the information so that pprof can visualize the call graphs. 

I am wondering if you have successfully profiled the heap on ARM using libgcc's backtrace. From #751, it sounds like you have. Perhaps I am mis-configuring it. I set these flags:

- ./configure --enable-stacktrace-via-backtrace --enable-emergency-malloc 

(although my assumption is that by default it will configure to use libgcc's s backtrace if libunwind is installed) Or I am just currently out of luck.

**Comment by alk** [comment]
Try running ./stacktrace_unittest with TCMALLOC_STACKTRACE_METHOD_VERBOSE set to 1.

E.g. 

$ TCMALLOC_STACKTRACE_METHOD_VERBOSE=1 ./stacktrace_unittest

It will tell you what it is using.

Also in order to get libgcc-enabled stacktrace capturing you don't need to pass --enable-stacktrace-via-backtrace. Just emergency malloc should be enough.

**Comment by vtk9** [comment]
Results below. 

```
TCMALLOC_STACKTRACE_METHOD_VERBOSE=1 ./stacktrace_unittest

Chosen stacktrace method is libgcc
Supported methods:
* libgcc
* generic
* arm

Obtained 8 stack frames.
Obtained 8 stack frames.
./stacktrace_unittest() [0x10df6] 0x10df6
./stacktrace_unittest() [0x11042] 0x11042
./stacktrace_unittest() [0x110fc] 0x110fc
./stacktrace_unittest() [0x111b8] 0x111b8
./stacktrace_unittest() [0x11274] 0x11274
./stacktrace_unittest() [0x11322] 0x11322
./stacktrace_unittest() [0x10bb8] 0x10bb8
/lib/arm-linux-gnueabihf/libc.so.6(__libc_start_main+0x9d) [0x76d7e8aa] 0x76d7e8aa
CheckStackTrace() addr: 0x112fd
Backtrace 0: expected: 0x10dec..0x10eda  actual: 0x10df6 ... OK
Backtrace 1: expected: 0x11038..0x11046  actual: 0x11042 ... OK
Backtrace 2: expected: 0x110f0..0x11100  actual: 0x110fc ... OK
Backtrace 3: expected: 0x111ac..0x111bc  actual: 0x111b8 ... OK
Backtrace 4: expected: 0x11268..0x11278  actual: 0x11274 ... OK
Backtrace 5: expected: 0x11314..0x11326  actual: 0x11322 ... OK
PASS
```
```

$ TCMALLOC_STACKTRACE_METHOD_VERBOSE=1 ./heap-checker_unittest.sh 
Testing ./heap-checker_unittest with HEAPCHECK= ... OK
Testing ./heap-checker_unittest with HEAPCHECK=local ... FAILED
Output from the failed run:
----
WARNING: Perftools heap leak checker is active -- Performance may suffer
Chosen stacktrace method is libgcc
Supported methods:
* libgcc
* generic
* arm



0x0 does not seem to point into code of function Mmapper at 0x13c1d! Stack frame collection must be off in MemoryRegionMap!Aborted (core dumped)
----
```

**Comment by alk** [comment]
ARM does have a number of failing unittests, but basic things still work.

So in your case, we do have working backtrace capturing. Which means your heap profiles should be working.

**Comment by vtk9** [comment]
Unfortunately, the heap profile does not have information needed by pprof. Below is a quick test I am trying to run to validate that heap profiling works. Do you have any other suggestions? I appreciate your help thus far. 

```
#include "gperftools/malloc_extension_c.h"
#include "gperftools/heap-profiler.h"
#include <stdlib.h>

void foo()
{
    malloc(12345);
    HeapProfilerDump("Reason");
}

int main()
{
    HeapProfilerStart("profile");
    if (IsHeapProfilerRunning() == 0) {
        return -1;
    }
    foo();
    foo();
    foo();
    foo();
    
    return 0;
}
```
`
g++ Hello.cxx -I$(GPERF_HOME)/src -L$(GPERF_HOME)/.libs -fno-builtin-malloc  -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -ltcmalloc`

```
cat profile.0003.heap 
heap profile:      3:    37035 [     3:    37035] @ heapprofile
     3:    37035 [     3:    37035] @

MAPPED_LIBRARIES:
....
```

**Comment by Umakumar01** [comment]
am facing the similar issue where heap profiler for ARM doesnt report the exact trace where memory leak is observed rather it shows only leak is observed

> heap profile:      6:      221 [     6:      221] @ heapprofile
     1:       33 [     1:       33] @
     1:       10 [     1:       10] @
     1:       10 [     1:       10] @
     1:       10 [     1:       10] @
     1:       10 [     1:       10] @
     1:      148 [     1:      148] @

MAPPED_LIBRARIES:
> any solution is available in gperftools 2.6.1 version

**Comment by alk** [comment]
Most recent everything is nearly perfect now. legacy 32 bit arms don't have working frame pointer stacktrace capturing but libgcc/libunwind work. And aarch64 is just pefect. All tests pass.
